isthmus/src/main/java/org/apache/calcite/jdbc/LookupCalciteSchema.java

```diff
@@ -0,0 +1,48 @@
+package org.apache.calcite.jdbc;
+
+import com.google.common.collect.Maps;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.impl.AbstractSchema;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class LookupCalciteSchema extends SimpleCalciteSchema {
+  private final Function<List<String>, Table> lookup;
+  private final Map<List<String>, Table> cache = Maps.newHashMap();
+
+  LookupCalciteSchema(
+      @Nullable CalciteSchema parent,
+      Schema schema,
+      String name,
+      Function<List<String>, Table> lookup) {
+    super(parent, schema, name);
+    this.lookup = lookup;
+  }
+
+  @Override
+  protected @Nullable CalciteSchema getImplicitSubSchema(String schemaName, boolean caseSensitive) {
+    if (cache.computeIfAbsent(path(schemaName), lookup) != null) {
```

Comment:
I don't know the impl of `path()`. Is this ensure that if we have `foo.bar.t1` and `bar.foo.t1`, nothing get confused?

Reply:
yes. Path returns the name of each subSchema plus the table name. This ensures that we are looking up a table in the backend by its full path